Support for JetBrains IdeaVim

ristomatti opened the pull request:
This adds preliminary support for IdeaVim.

**Progress:**

- [x] Support enough mappings for basic practical use
- [x] Include (comment) links to required plugins
- [x] Verify remaps to `helix.vim` mapping are actually needed
- [x] Verify included set commands align with Helix defaults
- [x] Update `README.md` for better discoverability

chtenb commented on src/helix.idea.vim:
```diff
@@ -0,0 +1,198 @@
+
+" Selection manipulation
+nmap <A-s> <Action>(EditorAddCaretPerSelectedLine)
+vmap <A-s> <Action>(EditorAddCaretPerSelectedLine)
```
Is there a specific reason you're using vmap instead of xmap here?

ristomatti replied:
No, it's a leftover from when I started the effort and was still unaware that `x*map` is for visual only and `v*map` affects also command. Fixed now.

chtenb commented on src/helix.idea.vim:
```diff
@@ -0,0 +1,198 @@
+xnoremap d dv
+
+" Movement
+unmap <Left>
```
Helix does support navigation with arrows

ristomatti replied:
It does but for some reason the mappings prevented selection with arrows. There's a few IdeaVim specific `set` options that seemed related (I have them on the default setting).

My intention with using `unmap` was to clear the remaps in `helix.vim`, but based on your comment, I assume this _should_ disable arrow completely? They do still work.

Would the correct way to do this be to do this?

```vim
map <Up> <Up>
map <Down> <Down>
map <Left> <Left>
map <Right> <Right>
```

chtenb replied:
Oh, I guess unmapping them then results in getting the native Idea behavior then. That's also fine of course, I was just confused to see this, because I assumed that the IdeaVim default bindings for arrows would just continue to work. They do for Vim and VsVim at least.

ristomatti replied:
Yep it feels confusing indeed. Yesterday I tried various ways with the `set` options but found no combination that would make the arrows behave just like hjkl when in visual mode. 

Perhaps there is something I didn't yet try, but even with this unmapping, I lose the selection if I switch from hjkl to arrows or vice versa.

chtenb replied:
Hm, you're right that extend mode doesn't work for arrow keys, because I neglected to map them in helix.vim

chtenb commented on src/helix.idea.vim:
```diff
@@ -0,0 +1,198 @@
+nmap <Space>C <Action>(CommentByBlockComment)
+xmap <Space>C <Action>(CommentByBlockComment)
+
+nmap <Space>y <Action>(EditorCopy)
```
Do the registers not work in this emulator?

ristomatti replied:
Yet another thing I'll need to test again. I did do that yesterday and ended up with this, but like mentioned on another comment, I better do it with a clean `helix.vim` config.

chtenb replied:
Are these remaps still necessary?

ristomatti replied:
Yep they are. I don't understand why though. Without these:
1. Copy text to clipboard from some other app, then paste with `<Space>p` works
2. Yank text from IDE using `<Space>y`, then do `:reg`, I see the contents are on register `"`. Paste to some other app -> nothing

Fidling with the `set` command to define registers could fix this but at least with just `helix.vim` loaded yanking does not work. `<Space>R` does not work either, but curiuosly `<Space>p` does???

chtenb replied:
That's weird, because I defined `xnoremap <Space>y "*yg`, so I would expect it to show in the `*` register, not the `"` register. The `*` is what Vim uses as system clipboard register. Anyway, if your current workaround works, it works :)

chtenb commented on src/helix.idea.vim:
```diff
@@ -0,0 +1,198 @@
+nmap <Space>g <Action>(RecentChangedFiles)
+xmap <Space>g <Action>(RecentChangedFiles)
+
+nmap <Space>G <Action>(Debug)
```
Maybe it's a good idea to gather all non-standard-Helix bindings in one spot, and comment it as such, so that people can easily see what's different from regular Helix

ristomatti replied:
I agree, I tried to make this as 1:1 to Helix defaults. I've got a separate config that overrides many mappings defined here. Debug seems to be default nowadays though:

![image](https://github.com/user-attachments/assets/55005235-94c6-45eb-b0a4-7ff9142829ee)

chtenb replied:
Ah, my version doesn't have this one yet :)

chtenb commented on src/helix.idea.vim:
```diff
@@ -0,0 +1,198 @@
+xnoremap * "ay:<C-u>let @/='<C-r>a'<CR>
+
+" View mode
+nmap zc <Action>(EditorScrollToCenter)
```
Is there a reason nmap is used instead of nnoremap? In general nnoremap should be the default, and nmap is used when you want a recursive binding.

ristomatti replied:
Thanks! I was under the impression `*noremap` means that the right hand side of the mapping will be handled with Vim defaults. I also saw many IdeaVim dotfiles use `map` and `nmap` all over the config.

I assume this could then break some mappings defined in `helix.vim`? :thinking:

chtenb replied:
In this case it probably makes no difference, since you're not actually referencing keys on the right hand side. But in general it's a good idea to use noremap by default. This way exceptions to the rule clearly stand out, which makes trouble shooting easier (in my experience, but it's also considered best practice in vim-land).

ristomatti replied:
It appears `noremap` mappings require using `:action FooBar<CR>` rather than the newer format `<Action>(Foobar)`. 

[IdeaVim docs](https://github.com/JetBrains/ideavim?tab=readme-ov-file#executing-actions) say this:

> Please don't use :action in mappings. Use <Action> instead.

No clue why that is, but perhaps the emulator is able to handle the actins more reliably/efficiently when given in this format. It's probably a safer bet to do that?

chtenb commented on src/helix.idea.vim:
```diff
@@ -0,0 +1,198 @@
+nmap C <Action>(EditorCloneCaretBelow)
+xmap C <Action>(EditorCloneCaretAbove)
+
+xnoremap ~ Ugv
```
I've noticed that you've introduced several substitutes for commands that are already defined in helix.vim. Do they not work? If so, I would put a comment that explains why it doesn't work (perhaps link to an issue in the emulator or something)

ristomatti replied:
Some did not seem to work otherwise, but I think I should go through each of these with an empty config and/or with just `helix.vim` applied. Most of the issues I've ran into were related to `helix.vim` mappings with `<Esc>` and/or mode switching. E.g. the selection disappears just before the command is applied.

Yesterday I fiddled especially with `~` and ``` ` ```. With these changes I get the expected result if the selection is done by an IDE action (e.g. `<Action>(EditorSelectWord)` = "expand selection") and after moving to word start/end with `h`/`l` then selecting the word with `w`/`b`. But if I do the same but explicitly change to visual mode before `w`/`b`, only the first or last letter changes case. So essentially `~`/``` ` ``` first cleared the selection.

ristomatti replied:
Actually now that I tried commenting the lines out, I get the exact same behavior. I guess I was trying to fix this issue but only at that point realized the issue with selection. :facepalm:

ristomatti replied:
It appears in each of the state that the action fails, prefixing it with `v` fixes the issue. The weird thing is that IdeaVim indicates visual mode is active even before pressing `v`. I think it's in a state of waiting if the previously pressed key is followed by another. There's an indicator in the statusbar that shows what you've typed so far. AFAIK it gets cleared when a mapping gets triggered.

ristomatti replied:
They're now removed, but I'll leave this unresolved to remember I need to check the others as well. Thanks!

chtenb reviewed (changes requested):
Nice job! I'm not an Idea user myself, so I can't test it out. But I will try my best to review the bindings.

chtenb commented:
Take your time to make changes. When you feel that you have a version that is good enough for public use, let me know and we can merge this.

ristomatti commented:
@chtenb I'll do that. Would it make sense for me to prefix the commits (and possible future ones)? For example `IdeaVim: Support x, y, z`?

chtenb commented:
Yeah, that sounds fine

ristomatti commented:
> Yeah, that sounds fine, but I don't care too much about the commit messages here

I was thinking about discoverability as well. It'd help spotting if there's relevant changes. Another reason is to simplify coming up with commit messages. Even if it won't matter to you, it does to me. :grin:

ristomatti commented:
I've added a checklist for tasks I'm planning to do before removing the draft status.

ristomatti commented:
Update: I've ditched the arrow keys unmaps (I'll leave them broken). I've also fixed `<C-w>` remaps, I had accidentally put also the following key within the brackets :facepalm:. I discovered that while a frustating debugging session trying to figure out why indent from `>` kept waiting for the next character.

ristomatti commented:
@chtenb I believe this is ok for review. The only thing left is to check is if these two `set` command are needed:

```vim
set matchit
set multiple-cursors
```

...these are IdeaVim's built in emulation of Vim plugins with the same name. With `argtextobj` I was able to support `maa` (match around argument).

Most of the remaps turned out unnecessary, but these seemed to behave more Helix-like in IdeaVim:

```vim
nnoremap e ve
nnoremap E vE
nnoremap w vw
nnoremap W vW
nnoremap b vb
nnoremap B vB

noremap X 0V
xnoremap X <nop>
```

chtenb commented on src/helix.idea.vim:
```diff
@@ -0,0 +1,189 @@
+xnoremap \= <Action>(ReformatCode)
+
+" Select/extend
+nnoremap e ve
```
Do the original helix.vim commands not work? I did some careful crafting on those to make such it would emulate the helix behavior for 99%. While I cannot test, I would imagine that these mappings are somewhat dissimilar to what helix does. Also, what if you press `e` twice, will that not cancel the selection because of the visual mode toggle?

ristomatti replied:
> Do the original helix.vim commands not work? I did some careful crafting on those to make such it would emulate the helix behavior for 99%.

I don't have the slightest clue what this actually does:
```vim
noremap e <Esc>maembviwovbviwvlmc`blmbhvl[`o`
```

... but it does indeed work perfectly on _Vim_. With IdeaVim sadly, this is how each movement behaves:

![recording-2024-10-04_21 47 18](https://github.com/user-attachments/assets/8d30240d-6fda-4ce8-8e7f-3807551c0211)

On first line, I repeat `e`, on second `w`, on third `b` and on fourth `B`.

> Also, what if you press e twice, will that not cancel the selection because of the visual mode toggle?

Yes I believe this is what happens. However, with these bindings, at least the first motion retains the selection. So they are more or less a workaround. The same motions with the included mappings:

![recording-2024-10-04_22 00 07](https://github.com/user-attachments/assets/c4044e94-a4ab-438c-a586-e16cfd786712)

ristomatti replied:
Is this good enough? My Vim-foo is not on the same level as yours. I don't have the slightest idea how to try to work around this...

chtenb replied:
I see, that looks like IdeaVim has some offset regressions. I  Although ````noremap e <Esc>maembviwovbviwvlmc`blmbhvl[`o```` might seem intimidating, my vim skills also aren't that good :) It's mostly a bunch of mark commands and recalls that I figured out using trial and error & the help documentation. But I totally understand if you want to keep the simple version.

ristomatti replied:
I'd be stubborn enough to try fixing it, but I know for certain it'd take hours and hours. I'm hoping someone comes up who has enough Vim background to hack it together from muscle memory. :D

chtenb commented on src/helix.idea.vim:
```diff
@@ -0,0 +1,189 @@
+"
+
+" repeat "vn" to select next search match
+xnoremap vn :action SelectNextOccurrence<CR>
```
What is the difference with `n`?

ristomatti replied:
If I search eg. "noremap", on normal mode it selects the next occurrence (as it should), but if I change to visual mode, `n` does exactly the same - it does not retain the previous selection. This allowed to add a cursor to next instance.

I did some more testing and this would behave exactly like in Helix in select mode:

```vim
xmap n <Action>(SelectNextOccurrence)
```

But it then breaks how `n` works in normal mode. To make things even more frustrating, I just noticed that the IDE default `F3` works in normal mode exactly as on Helix. It shows up as `FindNext` when I enable "IdeaVim: Track action ID's". I then thought this would work:

```vim
xmap n <Action>(SelectNextOccurrence)
nmap n <Action>(FindNext)
```

...but no. When it's called like this, `FindNext` activates visual mode, even though `F3` does not. :unamused:

I'll remove this mapping. `helix.vim` default works as expected on normal mode and I'll personally using `<C-d>` to add a cursor to the next match out of old habit from the times I used VSCode I believe. Heck, I even do that in Helix:

```toml
[keys.normal]
'C-d' = ["search_selection", "extend_search_next"]

[keys.select]
'C-d' = ["search_selection", "extend_search_next"]
```

ristomatti replied:
Removed.

chtenb commented on src/helix.idea.vim:
```diff
@@ -0,0 +1,189 @@
+nmap <Space>? <Action>(GotoAction)
+xmap <Space>? <Action>(GotoAction)
+
+nmap <Space>z <Action>(ToggleDistractionFreeMode)
```
This is non-standard, correct?

ristomatti replied:
Good catch! Removed.

ristomatti commented:
I've removed unused plugins and changed the `set` commands to align with Helix defaults AFAIK. No changes planned for this PR unless requested. :sweat_smile:

chtenb reviewed (approved):
Ok let's merge this. Subsequent improvements can go in a separate PR.

ristomatti commented:
Thanks for the thorough review!
